Feat/service

calebmer opened the pull request:
Not ready to be merged, but I wanted to start the review process. I've been working mainly in parallel with [`ardite-rest`](https://github.com/ardite/ardite-rest) adding stuff when/where it makes sense.

calebmer committed:
feat(service): add initial service struct

calebmer committed:
refactor(service): remove service struct

calebmer committed:
feat(driver): add memory driver

calebmer committed:
refactor(driver): only give driver type name

calebmer committed:
refactor(driver): optimize default read one method

calebmer committed:
doc(schema): add todo comment

calebmer committed:
fix(mongodb): fix errors

calebmer committed:
Revert "refactor(service): remove service struct"

This reverts commit af74dd7608796744302ad9668782e3636cc429dd.

calebmer committed:
feat(driver): add new method

calebmer committed:
feat(service): add some methods to service

calebmer committed:
feat(error): add not found method

calebmer committed:
fix(service): various compiler errors

calebmer committed:
refactor(driver): switch to not found error method

calebmer committed:
feat(service): add driver discovery

calebmer committed:
refactor(error): better error reporting

calebmer committed:
feat(service): add types function

calebmer committed:
refactor(schema): add some trait constraints

calebmer committed:
feat(error): add code to error object

calebmer committed:
style(*): prefer name-spacing over camel-spacing

It seems more idiomatic to write `error::Code` over `ErrorCode`. The want to write `ErrorCode` comes from my days in languages without namespaces, *cough* JavaScript *cough*.

calebmer committed:
refactor(*): reorganize exports

calebmer committed:
refactor(service): swap types method for definition getter

calebmer committed:
feat(service): and add it back

calebmer committed:
feat(value): add map keys function

calebmer committed:
feat(value): add a map value function

calebmer committed:
feat(case): add case enum

calebmer committed:
feat(value): add keys to case function

calebmer committed:
fix(case): add missing dependency

calebmer committed:
fix(value): silly errors

calebmer committed:
feat(case): wildcard re-export enum variants

calebmer committed:
feat(case): add is case method

calebmer committed:
refactor(schema): swap add_* with insert_*

calebmer committed:
feat(schema): inserted properties/types get converted into snake case

calebmer committed:
doc(case): add inline docs for case

calebmer committed:
refactor(case): use from str trait

calebmer committed:
feat(case): add more cases

calebmer committed:
refactor(case): remove from str trait

svmnotn commented on src/case.rs:
````diff
@@ -0,0 +1,146 @@
+  /// assert_eq!(Case::from_name("upper").unwrap(), Upper);
+  /// assert_eq!(Case::from_name("lower").unwrap(), Lower);
+  /// ```
+  pub fn from_name(string: &str) -> Option<Self> {
````
Just pointing out, if you only want lowercase here you can allow all sorts of weird mixes by using rust's [`String.to_lowercase()`](https://doc.rust-lang.org/std/string/struct.String.html#method.to_lowercase)

calebmer replied:
I know, but lower case is provided by the `Inflector` module, so I just use that implementation 😐

He might be using `to_lowercase` internally.

svmnotn replied:
I meant as, run the input to this through that so that it doesn't _need_ to be lower case

calebmer replied:
O. At this point I don't see this as incredibly necessary, and whoever is using this method could just run it themselves.

svmnotn commented on src/driver/discover.rs:
```diff
@@ -0,0 +1,24 @@
+// TODO: @svmnotn this is your jam!
+pub fn discover_driver(config: &schema::Driver) -> Result<Box<Driver>, Error> {
+  match config.url().scheme.as_str() {
+    "mongodb" => connect_mongodb_driver(config),
```
this will present a problem if the mongo feat. is not enabled.

calebmer replied:
Look at the definitions for `connect_mongodb_driver` a few lines down. It has two variants which conditionally compile depending on the feature.

svmnotn commented on src/driver/discover.rs:
```diff
@@ -0,0 +1,24 @@
+
+/// Takes a driver config value and finds and connects the associated driver.
+/// Errors if not driver could be found for the given config.
+// TODO: @svmnotn this is your jam!
```
Do you still want osx support?

calebmer replied:
Yes

calebmer replied:
Always

svmnotn commented on src/driver/driver.rs:
```diff
@@ -1,14 +1,13 @@
 /// interoperable with any data source, however the driver also assumes a
 /// collection based data model.
-pub trait Driver {
+pub trait Driver: Send + Sync {
```
why `Send + Sync`?

calebmer replied:
It's required by `iron` which spawns multiple threads. The driver type (used inside all these threads) needs to have these traits.

svmnotn commented on src/driver/memory.rs:
```diff
@@ -0,0 +1,31 @@
+use query::{Range, SortRule, Condition, Query};
+use value::{Key, Iter};
+
+pub struct Memory;
```
What is this?

calebmer replied:
The memory driver. If no other driver is specified, this one is used. I added docs so hopefully that clarifies things.

calebmer committed:
doc(*): add some inline docs

calebmer committed:
doc(case): add warning to inline docs

calebmer committed:
fix(error): fix inline example

calebmer committed:
fix(schema): remove unnesecary into

calebmer committed:
refactor(case): remove case in favor of new crate

calebmer committed:
feat(service): add driver functions to service

calebmer committed:
refactor(value): remove lifetime from iter

calebmer committed:
refactor(error): error expect function compiles out of tests

calebmer committed:
refactor(tests): move definition fixtures

calebmer committed:
fix(mongodb): fix errors

calebmer committed:
refactor(tests): rename tests

calebmer committed:
fix(mongodb): fix bad recursion error

calebmer committed:
test(mongodb): move tests into own module

calebmer committed:
feat(memory): add initial implementation of memory driver's read

calebmer committed:
refactor(driver): use &str and not &Key

calebmer committed:
feat(memory): implement range

calebmer committed:
feat(memory): implement condition testing

calebmer committed:
refactor(memory): use itertools for range

calebmer committed:
test(*): use git url for ardite driver tests

calebmer committed:
test(*): switch back to path…

ardite-driver-tests still depends on ardite-core at a certain path…

calebmer committed:
feat(query): add to pointers method

calebmer committed:
refactor(value): improve get function

calebmer committed:
test(memory): temporarily remove memory test

calebmer committed:
refactor(*): remove key and pointer

calebmer committed:
refactor(value): split into get and get path

calebmer committed:
refactor(schema): split into get and get path

calebmer committed:
feat(value): add set method

calebmer committed:
doc(value): add code examples

calebmer committed:
doc(*): replace pointer references with path

calebmer committed:
refactor(*): rename type to collection

calebmer committed:
Merge pull request #44 from ardite/refactor/no-key-or-pointer

Refactor/no key or pointer

calebmer committed:
refactor(query): remove query

calebmer committed:
Revert "test(memory): temporarily remove memory test"

This reverts commit 035039be3334bd5e4c6dada9dcbbea40fcc20477.

calebmer committed:
test(memory): remove linear map crate

calebmer committed:
feat(memory): implement sort

calebmer committed:
test(memory): remove memory driver tests

This removes the circular dependency. Memory driver tests will be in ardite-driver-tests

calebmer committed:
chore(mongodb): start fixing mongodb driver

calebmer committed:
fix(memory): rename to collection

calebmer committed:
fix(mongodb): fix driver

calebmer committed:
Merge pull request #43 from ardite/feat/memory

Feat/memory

calebmer committed:
refactor(service): dont use reference to internal driver lookup key

calebmer committed:
feat(service): add logging to driver connection

calebmer committed:
refactor(service): more efficient and performant

calebmer committed:
refactor(schema): remove like all of schema

calebmer committed:
refactor(error): modify error display

calebmer committed:
style(fixtures): capitalize type keys

calebmer committed:
doc(error): fix error example

calebmer committed:
feat(schema): add key property to type

calebmer committed:
refactor(value): use newtype pattern for value
